### DISCUSSION TITLE
SELECT * FROM read_parquet(SELECT filenames FROM table)

### DISCUSSION THREAD

**Discussion by DonIvanCorleone** [duckdb/duckdb#9019] [Q&A] [answered]
Hello there,

i currently try something potentially strange but for my use case necessary: 

I have a query on a table which gives me a list (0 to n elements) of locations of parquet files. I append the final parquet filename and try to create a list using something like this: `SELECT list(filepath||my_data.parquet) from table where column='important'`
 
Now I try to "stuff" this result into the following query `SELECT * from read_parquet(SELECT list(filepath||my_data.parquet) from table where column='important')`  which does not work neither does 

``` with ff as ( 
SELECT list(filepath||'my_data.parquet') from table where column='important')
 SELECT * from read_parquet(ff)
``` 

I am pretty convinced that I am doing something wrong but I cannot spot my mistake. Can anybody provide some help here? Many thanks in advance. 

Cheers

**Comment by l1t1** [comment]
```
copy (select 1 a) to 'a.parquet';
copy (select 2 a) to 'b.parquet';

create table a as from(values('a.parquet'),('b.parquet'))t(a);

select
'select * from read_parquet(["'  || array_to_string(list(a),'","')  ||  '"])'
from a;
```
then paste the result to CLI or use python to execute

**Reply by Tishj** [reply]
Yea, you'd need separate statements, we can already produce a list though, so it's less obtuse than the first example:
```sql
D insert into file VALUES('tmp/data.parquet');
D insert into file VALUES('tmp/data1.parquet');
D insert into file VALUES('tmp/data2.parquet');
D select * from file where not name.contains('1');
┌───────────────────┐
│       name        │
│      varchar      │
├───────────────────┤
│ tmp/data.parquet  │
│ tmp/data2.parquet │
└───────────────────┘
D select list(name) from file where not name.contains('1');
┌───────────────────────────────────────┐
│             list("name")              │
│               varchar[]               │
├───────────────────────────────────────┤
│ [tmp/data.parquet, tmp/data2.parquet] │
└───────────────────────────────────────┘
```

**Comment by DonIvanCorleone** [comment]
@Tishj @l1t1 
many thanks for your proposal, unfortunately, this does not work. I will try to explain a bit further what my needs are:

I am currently using the postgres extension to query a pg-database. This database only provides the path to the parquet. 
With some string concatenation I can manage to create the right file path. This result of query A needs to be plugged into query B which basically is `SELECT * FROM read_parquet(<result of query A>)` The important thing is: I do not have any access to python or CLI I am completely bound to the SQL editor.

Do you see another way to achieve my goals?

**Reply by DonIvanCorleone** [reply]
ADDENDUM: My current tries always lead me to the following error:

`duckdb error: Binder Error: Only table-in-out functions can have subquery parameters - read_parquet only accepts constant parameters`

**Reply by Tishj** [reply]
I don't see a way to achieve this without stepping out of the SQL editor.
Maybe in the future we will extend this to accept non-constant parameters, but I don't think there are plans for that yet

**Reply by davclark** [reply]
Thank you for providing this suggestion - it worked for me to shift to a python runtime and just concatenate my filenames into a list there. But definitely the need to hop into ipython or write a script is a bit less of the command-line warrior niche I think duckdb generally fills for me.

In particular, I think this kind of capability would be useful for doing development on data lake formats - it would allow me to directly query things in ducklake format, for example, and make sure I understood the logic, could try slightly different logic, etc. (query postgres for files + metadata, then convert that to a read_parquet() query).

So, if you do work on this capability, I'd love it!

**Reply by soerenwolfers** [reply]
this can now be done with variables, see the `read_csv` example at https://duckdb.org/docs/stable/sql/statements/set_variable.html

also, do consider upvoting my suggestion at https://github.com/duckdb/duckdb/discussions/17244 for a more elegant solution to this problem

**Comment by l1t1** [comment]
it's clear in the message above 'read_parquet only accepts constant'.
what you need is a method to execute a script, namely a.txt.  then you can fill the script with SQL such as copy (some query) to 'a.txt'

**Comment by l1t1** [comment]
consider the way in https://github.com/duckdb/duckdb/issues/9005, though it seems a bit silly.

**Comment by l1t1** [comment]
```
root@localhost:/data/data/com.termux/files/home/b/tbl.parquet/col=0# ls
data_0.parquet  data_2.parquet  data_4.parquet  data_6.parquet
data_1.parquet  data_3.parquet  data_5.parquet  data_7.parquet
root@localhost:/data/data/com.termux/files/home/b/tbl.parquet/col=0# /data/data/com.termux/files/home/b/duckdb
D select * from read_parquet('*',filename=true)where filename='data_0.parquet';
┌───────┬─────────┬────────────────┐
│  col  │    i    │    filename    │
│ int64 │  int64  │    varchar     │
├───────┼─────────┼────────────────┤
│     0 │       0 │ data_0.parquet │
│     0 │      10 │ data_0.parquet │
│     0 │      20 │ data_0.parquet │
│     0 │      30 │ data_0.parquet │

**Comment by l1t1** [comment, marked as answer]
```
D select * from read_parquet('*',filename=true)where substr(filename,6,1)in(1,3,6,8);
┌───────┬─────────┬────────────────┐
│  col  │    i    │    filename    │
│ int64 │  int64  │    varchar     │
├───────┼─────────┼────────────────┤
│     0 │  737280 │ data_6.parquet │
│     0 │  737290 │ data_6.parquet │
│     0 │  737300 │ data_6.parquet │

│     0 │ 9830370 │ data_3.parquet │
│     0 │ 9830380 │ data_3.parquet │
│     0 │ 9830390 │ data_3.parquet │
├───────┴─────────┴────────────────┤
│      405504 rows (40 shown)      │
└──────────────────────────────────┘
```

**Reply by DonIvanCorleone** [reply]
Many thanks for your support this was very helpful! After some testing I came to the same solution. I guess this is the best approach

**Comment by davetapley** [comment]
+1 for 
> accept non-constant parameters

I was trying to bootstrap a DB for testing from the [NYC taxi DB](https://www.nyc.gov/site/tlc/about/tlc-trip-record-data.page) with:

```sql
select * from read_parquet((select list(
    format('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_{}-{}.parquet',
    month(generate_series), year(generate_series)))
    from  generate_series(timestamp '2020-01-01', timestamp '2024-01-01', interval 1 month)));
```

But same 😞 

```
Error: Binder Error: Only table-in-out functions can have subquery parameters - read_parquet only accepts constant parameters
```

---

Workaround thanks to helpful Discord person who tipped me off to [nested functions](https://duckdb.org/docs/sql/functions/nested.html):
```sql
create table yellow as select * from
read_parquet(flatten(
[[('https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_{}-{:02d}.parquet').format(y, m)
for m in range(1, 13)]
for y in range(2021, 2024)]
));
```
